reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,270 @@
+
+#### What is the recommended minimal resource allocation of a Ray node?
+
+Each Ray node should have at least 4 CPU cores and 10GB of available memory. This corresponds to
+a Spark task configuration of:
+
+- ``spark.task.cpus >= 4`` and
+- ``(SPARK_WORKER_MEMORY / TASK_SLOTS_PER_SPARK_WORKER) >= 10GB``
+
+
+#### On a shared spark cluster, shall we let each user launch individual ray cluster, or one user can create a shared ray cluster and other user can also use it ?
+I think we can provide 2 API to support both cases.
+Shared mode:
+```ray.spark.get_or_create_shared_cluster```
+and
+```ray.spark.init_private_cluster```
+
+And if one shared mode ray cluster already exists, new ray cluster is not allowed to be launched.
+
+
+#### How to select the port number used by Ray node ?
+Ray node requires listening on several port, a spark cluster might be shared by many users,
+each users might setup their own ray cluster concurrently, to avoid port conflicts,
+we can randomly select free port and start Ray node service,
+if failed we can retry on another free port.
+
+
+#### How much memory shall we allocate to Ray node service ( set via ray script --memory option) ?
+Spark does not provide explicit API for getting task allowed memory,
+SPARK_WORKER_MEMORY / SPARK_WORKER_CORES * RAY_NODE_NUM_CPUS
+
+
+#### How will we allocate memory for the Ray object store?
+We will calculate the memory reserved for the Ray object store in each Ray node as follows:
+
+1. Calculate the available space in the ``/dev/shm`` mount.
+2. Divide the ``/dev/shm`` available space by the number of Spark tasks (Ray nodes) that can run
+   concurrently on a single Spark worker node, obtaining an upper bound on the per-Ray-node object
+   store memory allocation.
+3. To provide a buffer and guard against accidental overutilization of mount space, multiply the
+   upper bound from (2) by ``0.8``.
+
+This allocation will ensure that we do not exhaust ``/dev/shm`` storage and accidentally spill to
+disk, which would hurt performance substantially.
+
+
+#### How to make ray respect spark GPU resource scheduling ?
+In spark task, we can get GPU IDs allocated to this task, so, when launching
+Ray node, besides specifying `--num-gpus` options, we need to specify `CUDA_VISIBLE_DEVICES`
+environment so that we can restrict Ray node only uses the GPUs allocated to corresponding spark tasks.
+
+
+#### How to support all options of ray start
+Provide a ray_node_options argument (dict type).
+
+
+#### Where should the code live: ray repo or spark repo?
+Ray repo.
+Putting in spark repo is hard, It would be very hard to pass vote.
+Past examples includes horovod, petastorm, xgboost, tensorflow-spark, all of them tried to put in spark repo but failed.
+
+
+#### Does it support autoscaling ?
+We will be investigating and testing the feasibility of supporting autoscaling of ray worker nodes as 
+additional spark task slots become available upon the initiation of additional spark worker nodes.
```

Comment:
Non-blocking question, but could a spark expert perhaps point me to some code or docs about how "task slots" are managed?

Reply:
e.g., suppose a spark worker has 8 CPU cores, 4 GPU cores, and a spark task is configured by 2 cpus per task, and 1 gpu per task, then spark scheduler will calculate that the maximum task slots number is 4 on the spark worker.